从 0 创建自定义元素 (#5518)
xitu/gold-miner: TODO1/creating-a-custom-element-from-scratch.md

> * 原文作者：[Caleb Williams](https://css-tricks.com/author/calebdwilliams/)
> * 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> * 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/TODO1/creating-a-custom-element-from-scratch.md](https://github.com/xitu/gold-miner/blob/master/TODO1/creating-a-custom-element-from-scratch.md)

> > * 译者：ANFOUNNYSOUL

Reviewer: 这边放上link啦你肯定也希望大家都来看你的github 或者个人主页

> * 校对者：

# 从头创建自定义元素

> 在 [上一篇文章](https://github.com/xitu/gold-miner/blob/master/TODO1/crafting-reusable-html-templates.md), 我们通过创建文档中的HTML模板，但是我们想进行按需渲染，于是做了Web组件。

Reviewer: ```suggestion
在[上一篇文章](https://github.com/xitu/gold-miner/blob/master/TODO1/crafting-reusable-html-templates.md), 我们通过创建文档中的 HTML 模板，但是我们想进行按需渲染，于是做了 Web 组件。
```

> 在[上一篇文章](https://github.com/xitu/gold-miner/blob/master/TODO1/crafting-reusable-html-templates.md)，我们通过创建文档中的 HTML 模板，但是我们想进行按需渲染，于是做了 Web 组件。

Reviewer 2: 【我们通过创建文档中的 HTML 模板，但是我们想进行按需渲染，于是做了 Web 组件。】=>【我们在文档中创建了HTML模板，希望它们在需要时才呈现，这让我们开始接触Web组件】

Author: 感谢建议，请注意空格规范

> 接下来我们继续：下面有个目前仅使用 `HTMLTemplateElement` 的弹窗组件，我们会做一个自定义组件版本的

Reviewer: 弹窗组件 =》对话框组件

> 接下来我们继续：下面有个目前仅使用 `HTMLTemplateElement` 的对话框组件，我们会做一个自定义组件版本的。

Reviewer 2: 【接下来我们继续：下面有个目前仅使用 `HTMLTemplateElement` 的对话框组件，我们会做一个自定义组件版本的。】=>【接下来，我们将继续创建对话框组件的自定义元素版本，该自定义元素版本目前仅使用`HTMLTemplateElement`】

Reviewer 3: 这个我不同意。

定语从句 which currently only uses `HTMLTemplateElement` 的修饰对象应该是 the dialog component below。已经有的东西，才能说 currently 怎么怎么样。

我们要创建的，是 custom element version。要创建，也就是还没创建，所以不可能具有 currently only uses `HTMLTemplateElement` 的特性。

> See the Pen [Dialog with template with script](https://codepen.io/calebdwilliams/pen/JzjLyQ/) by Caleb Williams ([@calebdwilliams](https://codepen.io/calebdwilliams)) on [CodePen](https://codepen.io).

Reviewer: 这里是没有翻译吧

> 因此，我们继续创建一个自定义元素，该元素实时使用我们的`template#dialog-template`元素

Reviewer 3: push ahead by creating 指的是下一步要XXXX。译成“继续创建”会让人理解为 resume the process of creating

> 因此，我们下一步创建一个自定义元素，该元素实时使用我们的`template#dialog-template`元素

Reviewer: ```suggestion
因此，我们下一步创建一个自定义元素，该元素实时使用我们的 `template#dialog-template` 元素
```

element 上面翻译成元素，这里翻译成组件 需要保持一致。
我的意见是 元素节点

> #### 文章系列:

Reviewer: 需要和 后面一位译者商量一下这里

Author: 已经都改成系列文章了

> 1.  [**Web 组件简介** ](https://juejin.im/post/5c9a3cce5188252d9b3771ad)

Reviewer: ```suggestion
1.  [Web 组件简介](https://juejin.im/post/5c9a3cce5188252d9b3771ad)
```

> 2.  [**制作可复用的HTML模板**](https://github.com/xitu/gold-miner/blob/master/TODO1/crafting-reusable-html-templates.md)

Reviewer: ```suggestion
2.  [制作可复用的 HTML 模板](https://github.com/xitu/gold-miner/blob/master/TODO1/crafting-reusable-html-templates.md)
```

> 3.  [**从头开始创建自定义元素 (_本文_)**](https://github.com/xitu/gold-miner/blob/master/TODO1/creating-a-custom-element-from-scratch.md)

Reviewer: ```suggestion
3.  [从头开始创建自定义元素 (**本文**)](https://github.com/xitu/gold-miner/blob/master/TODO1/creating-a-custom-element-from-scratch.md)
```

> 4.  [**使用 Shadow DOM 封装样式和结构**](https://github.com/xitu/gold-miner/blob/master/TODO1/encapsulating-style-and-structure-with-shadow-dom.md)

Reviewer: ```suggestion
4.  [使用 Shadow DOM 封装样式和结构](https://github.com/xitu/gold-miner/blob/master/TODO1/encapsulating-style-and-structure-with-shadow-dom.md)
```

> 5.  [**Web 组件的高级工具**](https://github.com/xitu/gold-miner/blob/master/TODO1/advanced-tooling-for-web-components/.md)

Reviewer: ```suggestion
5.  [Web 组件的高级工具](https://github.com/xitu/gold-miner/blob/master/TODO1/advanced-tooling-for-web-components/.md)
```

* * *

### 添加一个自定义元素

>  bread 和 butter 都是 Web Components 的 **自定义元素**. 该`customElements` API 为我们提供了定义自定义HTML标签的路径，这些标签可以在包含定义类的任何文档中使用

Reviewer 3: bread and butter 是英文习语，指维生的基本物资

> Web 组件的基础元素是 **自定义元素**. 该`customElements` API 为我们提供了定义自定义HTML标签的路径，这些标签可以在包含定义类的任何文档中使用

Reviewer: ```suggestion
Web 组件的基础元素是**自定义元素**。该 `customElements` 的 API 为我们提供了定义自定义 HTML 标签的路径，这些标签可以在包含定义类的任何文档中使用。
```

> Web 组件的基础元素是**自定义元素**。该 `customElements` 的 API 为我们提供了定义自定义 HTML 标签的路径，这些标签可以在包含定义类的任何文档中使用。

Reviewer 2: 【为我们提供了定义自定义 HTML 标签的路径】=>【为我们提供了创建自定义 HTML 标签的途径】

> 可以把它想象成React 或 Angular 组件 (例如 `<MyCard />`), 不依赖React或Angular. 原生自定义组件是这样的: `<my-card></my-card>`. 更重要的是, 将它视为一个标准元素，可以在你的React，Angular，Vue, [insert-framework-you’re-interested-in-this-week] 应用中去使用，而不必大惊小怪。

Reviewer: ```suggestion
可以把它想象成 React 或 Angular 组件 (例如 `<MyCard />`)，不依赖 React 或 Angular。原生自定义组件是这样的：`<my-card></my-card>`。 更重要的是，将它视为一个标准元素，可以在你的React，Angular，Vue，[insert-framework-you’re-interested-in-this-week] 应用中去使用，而不必大惊小怪。
```

> 可以把它想象成React 或 Angular 组件 (例如 `<MyCard />`), 不依赖React或Angular. 原生自定义组件是这样的: `<my-card></my-card>`. 更重要的是, 将它视为一个标准元素，可以在你的React，Angular，Vue, [insert-framework-you’re-interested-in-this-week] 应用中去使用，而不必大惊小怪。

Reviewer: 这边我觉得得看一下原文 确定一下 [insert-framework-you’re-interested-in-this-week] 这个是连接掉了还是这么回事。

而不必大惊小怪 翻译太生硬了。我建议改语序和用词

Author: 链接原文就没有，我也考虑过大惊小怪，但是没有合适的插入语序，甚至可以放弃这句。

Reviewer: 嗯嗯 我也看了原文， 大概就是插入 任何你感兴趣的 应用之类的。 我的意思是翻译掉或者删掉也行。不要留着原文就好～

> 可以把它想象成React 或 Angular 组件 (例如 `<MyCard />`), 但没有React或Angular的依赖. 原生自定义组件是这样的: `<my-card></my-card>`. 更重要的是, 将它视为一个标准元素，可以在你的React，Angular，Vue, [insert-framework-you’re-interested-in-this-week] 应用中去使用，而不必大惊小怪。

Reviewer 3: 英文单词与前后的汉字间需要有1个半角空格

without the xxx dependency 就是指“不依赖XXX"，英文的词类不需要对译

> 可以把它想象成 React 或 Angular 组件（例如 `<MyCard />` ），不依赖React或Angular。原生自定义组件是这样的：`<my-card></my-card>`。更重要的是，将它视为一个标准元素，可以在你的 React、Angular、Vue、[insert-framework-you’re-interested-in-this-week]应用中去使用，而不必大惊小怪。

Reviewer 2: 【不依赖React或Angular】=>【但实际上它不依赖于 React 或 Angular】

> 从本质上讲, 一个自定义元素分为两个部分组成: 一个 **标签名称** 和一个 **Class** 类扩展内置  `HTMLElement` 类. 我们自定义元素的简易demo版本如下所示:

Reviewer: ```suggestion
从本质上讲, 一个自定义元素分为两个部分组成: 一个**标签名称**和一个 **Class** 类扩展内置 `HTMLElement` 类。我们自定义元素的简易 demo 版本如下所示:
```

customElements.define('one-dialog', OneDialog);
```

> 注意: 在整个自定义元素中, this值是对自身自定义元素实例的引用.

Reviewer: ```suggestion
注意：在整个自定义元素中， this 值是对自身自定义元素实例的引用。
```

Author: 我检查了一遍标点符号，没什么还要改的

> 在上面的示例中，我们定义了一个符合标准的新HTML元素，`<one-dialog></one-dialog>` 。它还做不了什么……暂时来说，在任何 HTML 文档中使用 `<one-dialog>` 标签将会创建一个带着 `<h1>` 标签显示 “Hello, World!” 的新元素。

Reviewer 2: 【它还做不了什么……暂时来说】=>【它暂时还做不了什么。现在，】

> 在上面的示例中, 我们定义了一个符合标准的新HTML元素, `<one-dialog></one-dialog>`. 它没有做太多事情... 但是现在, 在任何 HTML文章中使用 `<one-dialog>` 标签将会创建一个 `<h1>` 标签显示“Hello, World!” 的新元素

Reviewer 3: 英文单词与前后的汉字之间要有1个半角空格

It doesn’t do much… yet 指的是“它还做不了什么...暂时来说”

document 不是文章，是文档

with an `<h1>` tag 里面的 with an 不需要翻吗?

> 我们肯定想把它做的更 NB, 很幸运。 在[上一篇文章中](https://css-tricks.com/crafting-reusable-html-templates)，我们为弹出框创建模板，并且有权拿到模板，让我们在自定义元素中使用它。我们在该示例中添加了一个 script 标签来执行一些对话框魔术。我们暂时删除它，因为我们将把逻辑从 HTML 模板移到自定义元素类中。

Reviewer 2: 【并且有权拿到模板】=>【并且能够拿到模板】

> 我们肯定想把它做的更NB, 很幸运. 在 [上一篇文章中](https://css-tricks.com/crafting-reusable-html-templates), 我们为弹出框创建模板， 并且有权访问该模板 ，让我们在自定义元素中使用它。 我们在该示例中添加了一个script标签来执行一些对话框魔术。 我们暂时删除它，因为我们将把逻辑从HTML模板移到自定义元素类中。

Reviewer 3: will have access 这里的 access 不作 IT 用语的“访问权限”解，而仅仅是表示“能拿到手”。可以意译为“有XX可用”。

英文单词与前后的汉字间需要有 1 个半角空格

}
```

现在，定义了自定义元素 (`<one-dialog>`) 并指示浏览器呈现包含在调用自定义元素的HTML模板中的内容。

下一步是将我们的逻辑转移到组件类中。

### 自定义元素生命周期方法

> 与React或Angular一样，自定义元素具有 **生命周期方法**.  已经给你介绍过 `connectedCallback`, 当我们的元素被添加到DOM的时候调用它.

Reviewer 3: 英文单词与前后的汉字间需要有 1 个半角空格

英文为被动语态的句子，中文不能这样直接以动词开头。要 (1)补足主语、(2)补充主题语或者(3)取动词的被动义。这里可以用 (1) 我/笔者已经向你/您/各位XXXXX 或者 (3) XXXXX 已经向你/您/各位介绍过了。另外 passively 漏了

>  `connectedCallback` 与元素的 `constructor` 是分开的。函数用于设置元素的基本骨架，而 `connectedCallback` 通常用于向元素添加内容、设置事件监听器或以其他方式初始化组件。

Reviewer 2: 【建议保持对constructor翻译的一致性】

实际上, 构造函数不能用于设计或修改或操作元素的属性如果我们要使用创建对话框的新实例， `document.createElement`则会调用构造函数. 元素的使用者需要一个没有插入属性或内容的简单节点

> 该 `createElement` 函数没有配置将返回的元素的选项. 这是符合情理的, 那么话说回来了, 构造函数不应该有修改它创建的元素的能力.这使我们 `connectedCallback` 成为修改元素的地方.

Reviewer 3: [该 `createElement` 函数没有配置将返回的元素的选项.] => [该 `createElement` 函数没有可以用于配置将返回的元素的选项]

It stands to reason, then, that 里面 it 是形式主语，指示后面的动词不定式及从句才是主语。实际的语义是： to reason that constructor shouldn't xxxxxxx stands。意思是说，既然这个函数没有选项可以配置会返回的元素，那么我们就会合理地推想，这个函数应该也没XXXXXXX。

that leaves us with 意思其实就是说 then our only option is

> 在标准内置元素中，元素的状态通常通过元素上存在的属性和这些属性的值来反映。对于我们的示例，我们将仅查看一个属性：`[open]`。为此，我们需要关注该属性的更改，我们需要 `attributeChangedCallback` 来做到这一点。只要其中一个元素构造函数调用第二个生命周期方法 `observedAttributes` 更新。

Reviewer 2: 【我们需要关注该属性的更改】=>【我们需要观察该属性的变化】

> 在标准内置元素中，元素的状态通常通过元素上存在的属性和这些属性的值来反映。对于我们的示例，我们将仅查看一个属性：`[open]`。为此，我们需要关注该属性的更改，我们需要 `attributeChangedCallback` 来做到这一点。只要其中一个元素构造函数调用第二个生命周期方法 `observedAttributes` 更新。

Reviewer 2: 【只要其中一个元素构造函数调用第二个生命周期方法 `observedAttributes` 更新】=>【只要元素构造函数的 `observedAttributes` 之一发生变化，就会调用第二个生命周期方法。】

> 具有标准内置元素, 元素的状态通常通过元素上存在的属性和这些属性的值来反映。 对于我们的示例，我们将仅查看一个属性: `[open]`. 。为此，我们需要关注该属性的更改，我们需要`attributeChangedCallback` 要做到这一点. 只要其中一个元素构造函数调用第二个生命周期方法 `observedAttributes` 更新.

Reviewer 3: with 不是具有，而是指某种状况下。

}
```

> 在上面的例子中，我们只关心属性是否设置，我们不关心值（这类似于 HTML5 input 输入框上的 `required` 属性）。更新此属性时，我们更新元素的 `open` 属性。属性（property）存在于 JavaScript 对象上，HTML Elements 也具有属性（attribute）；这个生命周期方法可以帮助我们让两种属性保持同步。

Reviewer 2: 【我们只关心属性是否设置，我们不关心值】=>【我们只关心属性是否设置，并不关心具体的值】

我们将 updater 包含在 `attributeChangedCallback` 内部的条件检查中，以查看新值和旧值是否相等。我们这样做是为了防止程序中出现无限循环，因为稍后我们将创建一个 getter 和 setter 属性，它将通过在元素的属性（property）更新时设置元素的属性（attribute）来保持属性（attribute）和属性 （property）的同步。`attributeChangedCallback` 反向执行：当属性更改时更新属性。

> 现在，作者可以使用我们的组件，并且利用 `open` 属性决定对话框是否默认打开。为了使它更具动态性，我们可以在元素的 `open` 属性中添加自定义 getter 和 setter：

Reviewer 2: 【现在，作者可以使用我们的组件】=>【现在，开发者可以使用我们的组件】

}
```

getter 和 setter 将保证（HTML 元素节点上）的 `open` 特性和属性（在 DOM 对象上）的值同步。添加 `open` 特性会将 `element.open` 设置为 `true`，同理，将 `element.open` 设置为 `true` 会添加 `open` 属性。我们这样做是为了确保元素的状态由其属性反映出来。虽然在技术层面上不一定需要，但被认为是创建自定义元素的最优办法。

> 虽然这会不可避免地导致一些样板，但是通过循环观察到的属性列表并使用 `Object.defineProperty` 创建一个保持这些属性同步的抽象类是一项相当简单的任务。

Reviewer 2: 【虽然这会不可避免地导致一些样板】=>【虽然这难免引入一些样板文件】

  }
}

> // 在我们可以扩展抽象类，而不是直接扩展 HTMLElement

Reviewer 2: 【在我们可以扩展抽象类】=>【我们可以扩展抽象类】

class SomeElement extends AbstractClass { /** 省略 **/ }

customElements.define('some-element', SomeElement);
```

> 上面的例子并不完美，它没有考虑实现像 `open` 这样的属性的可能性，这些属性没有被分配给它们的值，而仅仅依赖于属性的存在。做一个完美的版本将超出本文的范围。

Reviewer 2: 【这些属性没有被分配给它们的值，而仅仅依赖于属性的存在】=>【这些属性没有被赋值，只是依赖于属性的存在】

}
```

> 这里有很多事情，但让我们来看看。我们要做的第一件事就是获取我们的容器，在 `isOpen` 的基础上切换 `.open` 类。为了使我们的元素可以访问，我们还需要切换 `aria-hidden` 属性

Reviewer 2: 【这里有很多事情，但让我们来看看。】=>【这里发生了很多事情，让我们来梳理一下。】

如果对话框已经打开了，那么我们希望保存对先前聚焦元素的引用。这是为了考虑可访问性标准。我们还将一个 keydown 监听器添加到名为 `WatEscape` 的文档中，该文档在构造函数中绑定元素的 `this`，其模式类似于 React 处理类组件中的方法调用的方式。

我们这样做不仅是为了确保正确绑定 `this.close`，还因为 `Function.prototype.bind` 返回带绑定调用栈的函数的实例。通过在构造函数中保存对新绑定方法的引用，我们可以在对话框断开时删除事件（稍后将详细介绍）。最后，我们将注意力集中在元素上，并将焦点设置在 shadow root 中的适当元素上。

> 我们还创建了一个很好的小实用工具方法来关闭我们的对话框，该对话框调度一个自定义事件，警告某些监听器已关闭对话框。

Reviewer 2: 【该对话框调度一个自定义事件，警告某些监听器已关闭对话框】=>【它分派一个自定义事件来通知某个监听器对话框已经关闭。】

如果元素是关闭的（即 `!open`），我们检查以确保 `this._wasFocused` 属性已定义并具有 `focus` 方法并调用该方法以将用户的焦点返回到常规 DOM。然后我们删除我们的事件监听器以避免任何内存泄漏。

> 说到为自己的代码做好清理善后，就自然也要说下我们采用了另一种生命周期方法：`disconnectedCallback`。`disconnectedCallback` 与 `connectedCallback` 相反，因为一旦元素从 DOM 中删除，就调用该方法，并允许我们清理绑定到元素的任何事件监听器或 `MutationObserver`。

Reviewer 2: 【因为一旦元素从 DOM 中删除，就调用该方法，并允许我们清理绑定到元素的任何事件监听器或 `MutationObserver`】=>【因为一旦从 DOM 中删除了元素，该方法就会被调用，它允许我们清理附加到元素的任何事件监听器或 `MutationObservers`。】

}
```

现在我们有一个运行良好，大部分可访问的对话框元素。我们可以做一些修饰，比如将焦点集中在元素上，但这超出了我们在本文学习的范围。

还有一个生命周期方法 `adoptedCallback`。它不适用于我们的元素，其作用是元素被采用（插入）到 DOM 的另一部分时触发。

在下面的示例中，您将看到我们的模板元素正被一个标准元素 `<one-dialog>` 所使用

请查阅笔记[对话框组件使用模板](https://codepen.io/calebdwilliams/pen/vbVXqv/) by Caleb Williams ([@calebdwilliams](https://codepen.io/calebdwilliams)) on [CodePen](https://codepen.io).

> ### 另一件事：非演示组件

Reviewer 2: 【另一件事】=>【另一个概念】
